pandas/io/parsers.py

```diff
@@ -786,17 +786,25 @@ def read(self, nrows=None):
             return ret
 
         # May alter columns / col_dict
-        index, columns, col_dict = self._create_index(ret)
+        index, columns, mappings, col_dict = self._create_index(ret)
 
-        df = DataFrame(col_dict, columns=columns, index=index)
+        if not mappings:
```

Comment:
Sure, the actual construction is separate, but getting to that construction requires some careful handling of the `names` and columns involved.  Also, performance is a major consideration.  The one thing I like about the current implementation (by which I mean, what is on `master`) is that it uses dictionaries, making indexing and updating more efficient.

Handling duplicates separately?  IMOH that's going to add a lot more (probably duplicate) code because you're going to have to provide similar functionality for two different cases.

Well, there are in fact loops already.  You loop through all of the elements in each spec provided and update the data and names according.  But one giant loop...not sure how feasible that is ATM.